DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs

```diff
@@ -154,6 +169,134 @@ public void VerifyHash_returns_false_for_hashes_created_outside_tolerance_time([
             result.Should().BeFalse();
         }
 
+        [Test]
+        public void ParseSsoAccountLinkingRequest_throws_exception_when_verifyHash_returns_false()
+        {
+            // Given
+            var linkLearningHubRequest = new LinkLearningHubRequest();
+
+            // When
+            Action act = () => learningHubSsoSecurityService.ParseSsoAccountLinkingRequest(linkLearningHubRequest);
+
+            // Then
+            act.Should().Throw<LearningHubLinkingRequestException>();
+        }
+
+        [Test]
+        public void ParseSsoAccountLinkingRequest_throws_exception_when_stored_sessionIdentifier_is_null()
+        {
+            // Given
+            var linkLearningHubRequest = new LinkLearningHubRequest
+            {
+                Hash = "vrdz",
+                State = $"{Guid.NewGuid()}_refId:1234",
+                UserId = 12345,
+            };
+
+            // When
+            Action act = () => learningHubSsoSecurityService.ParseSsoAccountLinkingRequest(linkLearningHubRequest);
+
+            // Then
+            act.Should().Throw<LearningHubLinkingRequestException>();
+        }
+
+        [Test]
+        public void ParseSsoAccountLinkingRequest_throws_exception_when_sessionIdentifier_does_not_match()
+        {
+            // Given
+            var linkLearningHubRequest = new LinkLearningHubRequest
+            {
+                Hash = "vrdz",
+                SessionIdentifier = Guid.NewGuid(),
+                State = $"{Guid.NewGuid()}_refId:1234",
+                UserId = 12345,
+            };
+
+            // When
+            Action act = () => learningHubSsoSecurityService.ParseSsoAccountLinkingRequest(linkLearningHubRequest);
+
+            // Then
+            act.Should().Throw<LearningHubLinkingRequestException>();
+        }
+
+        [Test]
+        public void ParseSsoAccountLinkingRequest_throws_exception_when_state_is_incomplete()
+        {
+            // Given
+            var linkLearningHubRequest = new LinkLearningHubRequest
+            {
+                Hash = "vrdz",
+                SessionIdentifier = Guid.NewGuid(),

```

Comment:
If I understand this correctly, this setup would still fail if the state was complete because the two sessionIdentifiers will differ. I think it would be better if each test request was as valid as possible save for the one error. This may also apply to the `verifyHash_returns_false` test.

Reply:
Although both of these tests do "fail" for the right reason I get your point so I have updated them ensuring the input data represents more clearly what is being tested for.